lte/gateway/python/magma/pipelined/app/apn.py

```diff
@@ -0,0 +1,284 @@
+
+from ryu.controller.controller import  Datapath
+from ryu.ofproto import nicira_ext
+from ryu.ofproto import ether
+from ryu.controller.handler import MAIN_DISPATCHER, set_ev_cls
+from ryu.controller import ofp_event
+
+
+class APNController(MagmaController):
+    """
+    APN Controller.
+
+    The APN controller is responsible for marking a flow with an encoded APN 
+    name. The APN name should be stored in
+    register.
+
+    """
+
+    APP_NAME = "apn"
+    APP_TYPE = ControllerType.LOGICAL
+
+    def __init__(self, *args, **kwargs):
+        super(APNController, self).__init__(*args, **kwargs)
+        # TODO(119vik) get following from the config file
+        self._clean_start = True 
+        self._datapath = None
+        self.tbl_num = self._service_manager.get_table_num(self.APP_NAME)
+        self.apn_tagging_scratch = \
+            self._service_manager.allocate_scratch_tables(self.APP_NAME, 1)[0]
+        self.next_table = self._service_manager.get_next_table_num(
+            self.APP_NAME)
+        self._msg_hub = MessageHub(self.logger)
+
+    def initialize_on_connect(self, datapath):
+        """
+        Install the default flows on datapath connect event.
+
+        Args:
+            datapath: ryu datapath struct
+        """
+        self._datapath = datapath
+        # In case wee need to clean all existing  buggy / orphaned flows 
+        # before start the controller
+        if self._clean_start:
+          self.delete_all_flows(datapath)
+
+    @set_ev_cls(ofp_event.EventOFPBarrierReply, MAIN_DISPATCHER)
+    def _handle_barrier(self, ev):
+        self._msg_hub.handle_barrier(ev)
+
+    @set_ev_cls(ofp_event.EventOFPErrorMsg, MAIN_DISPATCHER)
+    def _handle_error(self, ev):
+        self._msg_hub.handle_error(ev)
+
+    def _wait_for_flow_responses(self, imsi, flow_list, chan):
+        def fail(err):
+            self.logger.error(
+                "Failed to install flow for subscriber %s: %s",
+                imsi, err)
+            return False
+
+        for _ in range(len(flow_list)):
+            try:
+                result = chan.get()
+            except MsgChannel.Timeout:
+                return fail("No response from OVS")
+            if not result.ok():
+                return fail(result.exception())
+        return True
+
+    def cleanup_on_disconnect(self, datapath):
+        """
+        Cleanup flows on datapath disconnect event.
+
+        Args:
+            datapath: ryu datapath struct
+        """
+        self.delete_all_flows(datapath)
+
+    def delete_all_flows(self, datapath):
+        """Delete all flows which set APN register"""
+        flows.delete_all_flows_from_table(datapath, self.tbl_num)
+        flows.delete_all_flows_from_table(datapath, self.apn_tagging_scratch)
+
+    def add_apn_flow_for_ue(self, imsi, ue_ip_addr, apn):
+        """ 
+            Add flow which match all IN traffic with specified UE_IP and set 
```

Comment:
[nit] move whats inside the docstring one tab to the left

Reply:
fixed